### ISSUE TITLE
Use debug: option or use RACK_ENV to controller precompile process.

### ISSUE THREAD

**Issue by zw963** [hmdne/roda-sprockets#10]
if user provide both `debug options` + `RACK_ENV`, we should respect which one first?

For now, if test with `debug: false` + `RACK_ENV=development` rake assets:precompile, there is no error raised,
but will generate minified pipeline (it perfer to use debug settings)

And `debug: true` + `RACK_ENV=production`, assets not minified.

That is, we always prefer to use debug: option, this is acceptable anyway, though i propose should pefer use RACK_ENV in command line over debug: option.

Anyway, we should let  user to know, even, when use run task like
`RACK_ENV=production rake assets:precompile` on production, it still generate non-minified assets.

Anyway, i consider this is a issue which need discuss.

**Comment by hmdne** [comment]
I think RACK_ENV mostly an application implementer's detail - even though RACK_ENV is also used in rack quite universally, though our `debug:` variable could default to this. I don't know, maybe you have a better idea

In the recent commit I added a possibility to precompile the assets without a need for setting RACK_ENV as I believe this is what the users should expect.

**Comment by hmdne** [comment]
Also, doing it with `debug:` makes this thing testable.

**Comment by zw963** [comment]
> In the recent commit I added a possibility to precompile the assets without a need for setting RACK_ENV as I believe this is what the users should expect.

Yes, acceptable, but, anyway, i thought we need mention this in document, because for assets precompile, user expect different behavior when in devlopment/production, many user may assume
this behavior controll by envs, so, maybe user don't set debug: option, that will cause assets never be minified.

**Comment by hmdne** [comment]
This is the reason for the issue:

https://github.com/zw963/marketbet_crawler/blob/master/app/app.rb#L17

**Comment by hmdne** [comment]
After I will release v1.1.0, I will release v2.0.0 with an incompatible change that sets `debug` to an appropriate value by default.

**Comment by zw963** [comment]
Okay! before release v1.1.0, double confirm, path_path set to `public/assets/` instead of `public/assets` is intended, latter will cause issue, right? 

![image](https://user-images.githubusercontent.com/549126/130383286-9b097ae4-6a91-4821-aa78-950743d90ccc.png)

**Comment by hmdne** [comment]
This shouldn't matter at all. `prefix` treats / specially (I will document this in a moment), but here no.

**Comment by zw963** [comment]
I test on newest master:

If we never set debug: option

- RACK_ENV=production rake assets:precompile will unify assets
- rake assets:precompile not

If we set debug: option:, debug option always win.
- debug: false will unify assets
- debug: true will not.

this should be expected behavior we want to archive, thank you.

**Comment by hmdne** [comment]
Here I must disagree though. And I will answer why:

- User should not think about all this, just execute `rake assets:precompile` and get the best result (and also, while I don't like to use that argument, big frameworks like Rails work the same)
- I see no reason to build non-minified files when minify is enabled (other than to debug, but then the name `precompile` makes it obvious it's for production mode)
- an argument to `js_compressor:` can also be conditional
- RACK_ENV in general is an environment variable for Rack, and while running Rake tasks we don't really interface Rack in any way
- from what I know about Sprockets, the `debug` pipeline doesn't support precompiling at all, the previous behavior compiled files for the `default` (ie. production) pipeline, we just didn't set some variables (ie. minification)
- you still don't get source maps, because `default` doesn't produce them (now I checked and it can produce private source maps though - which is a very nice thing, but Sprockets support for that is buggy)

If you have some usecase I don't see, please feel free do describe it, but I'm quite sure this behavior you describe is not how it should really work. Maybe this is an issue: https://github.com/hmdne/roda-sprockets/blob/master/lib/roda/plugins/sprockets.rb#L41 (the `key?` check), because probably for Rake this ENV variable isn't even set

**Comment by hmdne** [comment]
I think this is the best solution for now:
- debug = true, only when RACK_ENV==development, so false in rake tasks and in production/test/staging/etc.
- debug being set overrides all other behavior
- code cleanup, so we become a thin layer after Sprockets and user can modify it as he wishes

**Comment by zw963** [comment]
>     * debug = true, only when RACK_ENV==development, so false in rake tasks and in production/test/staging/etc.

Agree with you.

> because probably for Rake this ENV variable isn't even set

Yes, though, i consider user responsible for setting the right environment, and, in fact, in production, at least for me, RACK_ENV was set globally in ~/.bashrc or /etc/profile,

**Comment by zw963** [comment]
> I think this is the best solution for now:

Sorry for confusing (again), because i set a RACK_ENV default value on my own app, this misleading me, Following is the newest manual test result on 2.0.0RC1

### debug: option not set

1. rake assets:precompile => minified
2. RACK_ENV=production rake assets:precompile => minified
3.  RACK_ENV=development rake assets:precompile => not

### debug: true
RACK_ENV=production rake assets:precompile => not

### debug: false
RACK_ENV=development rake assets:precompile => minified

**Comment by zw963** [comment]
I have some tests, for test the layout page. (though, this is maybe not so necessary)

it failed because, test use `RACK_ENV=test` by default.

```diff
--- expected
+++ actual
@@ -5,7 +5,7 @@
     <meta name=\"viewport\" content=\"width=device-width, initial-scale=1, shrink-to-fit=no\">
     <title>  Stock list
 </title>
-    <link rel=\"stylesheet\" type=\"text/css\" href=\"/assets/app.debug.css\">
+    <link rel=\"stylesheet\" type=\"text/css\" href=\"/assets/app-229e34816dcaa860ccefb78c1330f69e9711c10bc3d8c6b7afa8d2190618d249.css\">
   </head>
   <body>
     <div id=\"app\">
@@ -40,7 +40,7 @@
 
 
 
-    <script src=\"/assets/app.debug.js\" type=\"text/javascript\"></script>
+    <script src=\"/assets/app-f119d28b5edf6d5209073ddc1bf9e3ddbd415cc739cc608b3f7fc5a97e0d1c90.js\" type=\"text/javascript\"></script>
     <script> Opal.loaded(typeof(OpalLoaded) === \"undefined\" ? [] : OpalLoaded); Opal.require(\"app\"); </script>
       We have 2 stocks.
 
```

**Comment by zw963** [comment]
Another issue maybe need to mention, after use roda-sprockets plugins, when run rake test with `rack-test` gem, it very slow before access `/stocks` path, which use layout.erb,  which add tags like following:

```erb
<%== stylesheet_tag 'app' %>
    <%== javascript_tag 'app' %>
    <%== opal_require 'app' %>
```

It's seem like spent 98.73% time on `Kernel.system`, so, i consider it is compiling asserts, right?

Because i add following code into app.

```rb
r.sprockets if ENV['RACK_ENV'] == 'development'
```

so, `r.sprockets` should never by run anyway, because my test RACK_ENV == 'test'

I am no idea why still so slow.

Following is generated profile html file.

[result.html.gz](https://github.com/hmdne/roda-sprockets/files/7031723/result.html.gz)



Thank you.

**Comment by hmdne** [comment]
To speed up, you can either precompile before starting a test or enable cache. The fact that you disable the sprockets route does not mean that sprockets does not compile, because you call it via helpers as well.

**Comment by hmdne** [comment]
> I have some tests, for test the layout page. (though, this is maybe not so necessary)
> 
> it failed because, test use `RACK_ENV=test` by default.
> 
> ```diff
> --- expected
> +++ actual
> @@ -5,7 +5,7 @@
>      <meta name=\"viewport\" content=\"width=device-width, initial-scale=1, shrink-to-fit=no\">
>      <title>  Stock list
>  </title>
> -    <link rel=\"stylesheet\" type=\"text/css\" href=\"/assets/app.debug.css\">
> +    <link rel=\"stylesheet\" type=\"text/css\" href=\"/assets/app-229e34816dcaa860ccefb78c1330f69e9711c10bc3d8c6b7afa8d2190618d249.css\">
>    </head>
>    <body>
>      <div id=\"app\">
> @@ -40,7 +40,7 @@
>  
>  
>  
> -    <script src=\"/assets/app.debug.js\" type=\"text/javascript\"></script>
> +    <script src=\"/assets/app-f119d28b5edf6d5209073ddc1bf9e3ddbd415cc739cc608b3f7fc5a97e0d1c90.js\" type=\"text/javascript\"></script>
>      <script> Opal.loaded(typeof(OpalLoaded) === \"undefined\" ? [] : OpalLoaded); Opal.require(\"app\"); </script>
>        We have 2 stocks.
>  
> ```

The diff comes from the fact, that in RACK_ENV='test' debug is false by default

**Comment by zw963** [comment]
> The diff comes from the fact, that in RACK_ENV='test' debug is false by default

Yes, this is the concern, unless i don't want write page test, if i want, i have to test base on those precompiled (name suffix with digest) script tag, i don't know if this is a good idea to fixing test when assets digest was changed.

**Comment by hmdne** [comment]
Ok, that's a valid concern. In general, RACK_ENV shouldn't even be used for values other than `development` and `deployment`, but it is used and I'm concerned about values such as `staging`. The next commit will treat RACK_ENV='test' just as 'development'.

**Comment by zw963** [comment]
It works!
